File: double pendulum simulation (RK4)/Cpp version/DoublePendulum/Source.cpp
/*
 =============================================================================
 Simulating the double pendulum using Runge–Kutta method(RK4)
 =============================================================================

Updated on Mon Jul 20 2020
@author: Ahmed Alkharusi*/





#include <iostream>
#include <vector>
#include <cmath>
#include <fstream>
#include <string>
#include "Array.h"


const double m1{ 1.0 }; //mass of the 1st pendulum 
const double m2{ 1.0 }; //mass of the 2nd pendulum 
const double g{ 10.0 }; //gravity
const double r1{ 1.0 }; //length of the 1st pendulum   
const double r2{ 1.0 }; //length of the 2nd pendulum   
const double pi{ 3.14159265359 };



Array deriv_a1(Array a1_arr, Array a2_arr, Array t);
Array deriv_a2(Array a2_arr, Array a1_arr, Array t);
void rk4(Array(*deriv)(Array, Array, Array), Array &func_i, Array &func_i2, double &x_i, const double h);
void output_matrix(std::vector<std::vector<double>> matrix, std::string&& file_name);


int main() {
	
	const double step_size{ 0.001 };//step size for the RK4 method
	const long unsigned int steps_no { 6000000 };//number of steps of the RK4 method
	//initial conditions
	double t{ 0 };
	Array a1_arr{ pi/2 , -3 };
	Array a2_arr{ pi/2 , 3 };

	std::vector<double> time_arr{ t };
	std::vector<std::vector<double>> Pendulum1_results;
	std::vector<std::vector<double>> Pendulum2_results;
	
	std::vector<double> temp_Pendulum2_results{t, a2_arr.angle};
	std::vector<double> temp_Pendulum1_results{t, a1_arr.angle};

	Pendulum2_results.push_back(temp_Pendulum2_results);
	Pendulum1_results.push_back(temp_Pendulum1_results);
	
	
	for (size_t i{ 0 }; i < steps_no;i++) {
		Array temp = a1_arr;
		rk4(&deriv_a1, a1_arr, a2_arr, t, step_size);
		t -= step_size;
		rk4(&deriv_a2, a2_arr, temp, t, step_size);
		temp_Pendulum2_results[0]= temp_Pendulum1_results[0]= t;
		temp_Pendulum2_results[1] = a2_arr.angle; // can output the angular speed by changing this from .angle -> .angular_speed
		temp_Pendulum1_results[1] = a1_arr.angle; // can output the angular speed by changing this from .angle -> .angular_speed
		Pendulum2_results.push_back(temp_Pendulum2_results);
		Pendulum1_results.push_back(temp_Pendulum1_results);
	}
	output_matrix(Pendulum1_results,"Pendulum1_results.txt"); //outputs to cols. 1st is the time, 2nd is the angle 
	output_matrix(Pendulum2_results, "Pendulum2_results.txt");
	return 0;
}

Array deriv_a1(Array a1_arr, Array a2_arr, Array t) {
	double num = -g * (2 * m1 + m2) * std::sin(a1_arr.angle) - m2 * g * std::sin(a1_arr.angle - 2 * a2_arr.angle) - 2 * m2 * std::sin(a1_arr.angle - a2_arr.angle) * (r2 * std::pow(a2_arr.angular_speed, 2) + r1 * std::pow(a1_arr.angular_speed, 2) * std::cos(a1_arr.angle - a2_arr.angle));
	double den = r1 * (2 * m1 + m2 - m2 * std::cos(2 * a1_arr.angle - 2 * a2_arr.angle));
	// (num / den) is the angular speed
	Array temp{ a1_arr.angular_speed,num/den };
	return  temp;
}
Array deriv_a2(Array a2_arr, Array a1_arr, Array t) { 
	double temp = (2 * std::sin(a1_arr.angle - a2_arr.angle));
	double num = temp * (r1 * std::pow(a1_arr.angular_speed, 2) * (m1 + m2) + g * (m1 + m2) * std::cos(a1_arr.angle) + r2 * std::pow(a2_arr.angular_speed, 2) * m2 * std::cos(a1_arr.angle - a2_arr.angle));
	double den = r2 * (2 * m1 + m2 - m2 * std::cos(2 * a1_arr.angle - 2 * a2_arr.angle));
	// (num / den) is the angular speed
	Array res{ a2_arr.angular_speed, num / den };
	return  res;
}

void rk4(Array(*deriv)(Array, Array, Array), Array& func_i, Array& func_i2, double& x_i, const double h) {
	Array k1 = deriv(func_i, func_i2, (k1*0)+x_i);
	Array k2 = deriv(func_i + h / 2, func_i2,  k1 * (h/ 2) );
	Array k3 = deriv(func_i + h / 2, func_i2, k2 * (h / 2));
	Array k4 = deriv(func_i + h, func_i2, k3 * h);
	func_i = func_i + ((k1 + (k2 * 2) + (k3 * 2) + k4)) * (h / 6);
	x_i += h;
}
void output_matrix(std::vector<std::vector<double>> matrix, std::string &&file_name) {
	std::ofstream output_file(file_name);
	for (size_t i{ 0 };i < matrix.size();i++) {
		for (size_t j{ 0 };j < (matrix.at(i)).size();j++) {
			output_file << (matrix.at(i)).at(j) << " ";
		}
		output_file << "\n";
	}
}
/*
 =============================================================================
 Please check the answers!!!
 =============================================================================
References:

#Implementing the RK4 method in Python 
https://youtu.be/mqoqAovXxWA
by Prof. Niels Walet  

#The formulas for the angular acceleration 
https://www.myphysicslab.com/pendulum/double-pendulum-en.html
    
#Animating the double pendulum (N.B. the implementation used here is different)
https://matplotlib.org/3.2.1/gallery/animation/double_pendulum_sgskip.html

*/

File: double pendulum simulation (RK4)/Source.py
"""
# =============================================================================
# Simulating the double pendulum using Runge–Kutta method (RK4)
# =============================================================================

Created on Fri Jul 17 2020
@author: Ahmed Alkharusi
"""
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import animation
# =============================================================================
# globals
# =============================================================================
m1 = 1   #mass of the 1st pendulum 
m2 = 1    #mass of the 2nd pendulum   
g = 10 #gravity
r1 = 1  #length of the 1st pendulum   
r2 = 1   #length of the 2nd pendulum
x = y = []
# =============================================================================
# Functions defn.
# =============================================================================

def angular_acc1(a1_arr,a2_arr):
    """Calculate the angular acceleration for the 1st pendulum
        Inputs-> a1_arr: np.array([Initial angle, Initial angular velocity]);
                 a2_arr: np.array([Initial angle, Initial angular velocity]);"""
    num = -g *(2*m1+m2)*np.sin(a1_arr[0]) - m2*g*np.sin(a1_arr[0]-2*a2_arr[0])- 2* m2*np.sin(a1_arr[0]-a2_arr[0]) * (r2*pow(a2_arr[1],2)+r1*pow(a1_arr[1],2)*np.cos(a1_arr[0]-a2_arr[0]))
    den = r1*(2*m1+m2-m2 * np.cos(2*a1_arr[0]-2*a2_arr[0]))
    return num/den

def angular_acc2(a1_arr,a2_arr):
    """Calculate the angular acceleration for the 2nd pendulum
        Inputs-> a1_arr: np.array([Initial angle, Initial angular velocity]);
                 a2_arr: np.array([Initial angle, Initial angular velocity]);"""
    temp = (2*np.sin(a1_arr[0]-a2_arr[0])) 
    num = temp * (r1*pow(a1_arr[1],2)*(m1+m2)+g*(m1+m2)*np.cos(a1_arr[0])+r2*pow(a2_arr[1],2)*m2*np.cos(a1_arr[0]-a2_arr[0]))
    den = r2*(2*m1+m2-m2 * np.cos(2*a1_arr[0]-2*a2_arr[0]))
    return num/den

def deriv_a1(a1_arr,a2_arr,t):
    """
    Returns an array np.array([first derivative, 2nd derivative])
    Inputs-> a1_arr: np.array([Initial angle, Initial angular velocity]);
             a2_arr: np.array([Initial angle, Initial angular velocity]);
             t: the dependent variable;
    """
    return np.array([a1_arr[1],angular_acc1(a1_arr,a2_arr)])

def deriv_a2(a2_arr,a1_arr,t):
    return np.array([a2_arr[1],angular_acc2(a1_arr,a2_arr)])

def rk4(deriv,func_i,func_i2, x_i,h):
    """
    Implements the RK4 method
    Inputs-> deriv: a function that takes two arguments;
             func_i: the function to be calculated;
             func_i2: this is just passed as an argument for func_i (see above deriv_a1 and deriv_a2);
             x_i: the dependent variable of func_i;
             h: the step size;
             
    """
    k1 = deriv(func_i,func_i2,x_i)
    k2 = deriv(func_i+h/2,func_i2,h*k1/2)
    k3 = deriv(func_i+h/2,func_i2,h*k2/2)
    k4 = deriv(func_i+h,func_i2,h*k3)
    func = func_i + (1/6) * h * (k1 +2*k2+2*k3+k4)
    x = x_i + h
    return (x,func)

# =============================================================================
# def init(): #Uncomment these for the animation
#     line.set_data([], [])
#     time_text.set_text('')
#     return line, time_text
# 
# def animate(i):
#     x = [0, pendulum1_x[i], pendulum2_x[i]]
#     y = [0, pendulum1_y[i], pendulum2_y[i]]
#     
#     line.set_data(x,y)
#     #time_text.set_text(time_template % (i*h)) #Uncomment this to display the time in the animated plot
#     return line, time_text
# 
# =============================================================================
# =============================================================================
# The results
# =============================================================================
    
#Initial conditions ([initial angle, initail angular speed])
a1_arr = np.array([np.pi/2,0])
a2_arr = np.array([np.pi/2,1])
t = 0 # starting time
h = 0.001 # step size for the RK4 method
steps_no = 100000 # number of steps of the RK4 method
time_arr = np.array([t])
func_array1 = np.array([a1_arr])
func_array2 = np.array([a2_arr])

for i in range(steps_no):
    temp =a1_arr
    (t,a1_arr) = rk4(deriv_a1,a1_arr,a2_arr,t,h)
    t =-h 
    (t,a2_arr) = rk4(deriv_a2,a2_arr,temp,t,h)
    time_arr2 = np.append(time_arr, t)
    func_array1 = np.vstack((func_array1,np.array([a1_arr])))
    func_array2 = np.vstack((func_array2,np.array([a2_arr])))
 
# You can plot the pendulum's position or angular speed/acceleration as a function of time
[pendulum1_theta, pendulum1_angular_speed] = func_array1.transpose()
[pendulum2_theta, pendulum2_angular_speed] = func_array2.transpose()

pendulum1_x = r1*np.sin(pendulum1_theta)
pendulum1_y = - r1*np.cos(pendulum1_theta)

pendulum2_x = r2*np.sin(pendulum2_theta) + pendulum1_x
pendulum2_y = pendulum1_y - r2*np.cos(pendulum2_theta)


# Here I used the matplotlib template of the double pendulum animation to animate the plot
# =============================================================================
# fig = plt.figure()
# ax = fig.add_subplot(111, autoscale_on=False, xlim=(-3.9, 3.9), ylim=(-2, 2))
# ax.set_xlabel('$x-Axis$',fontsize=12)
# ax.set_ylabel('$y-Axis$',fontsize=12)
# ax.set_title('Double pendulum simulation (RK4 method)',fontsize=14)
# ax.grid()
# 
# line, = ax.plot([], [], 'o-',lw=3,color='mediumvioletred',markersize=15)
# time_template = 'time = %0.1fs'
# time_text = ax.text(0.05, 0.9, '', transform=ax.transAxes)
# 
# ani = animation.FuncAnimation(fig, animate, np.arange(1, len(pendulum1_y)),
#                               interval=0, blit=True, init_func=init)
# 
# #ax.scatter(pendulum2_x, pendulum2_y,s=5, color='black',alpha=0.5)
# #ani.save('double_pendulum_200.avi', fps=20, dpi =8)
# plt.show()
# =============================================================================
# =============================================================================
# #Save each frame separately
# =============================================================================
scatter_x = []
scatter_y = []
counter = 0
save_every_n_frames = 25
for j in range(int(len(pendulum1_y)/save_every_n_frames)):
    i = j*save_every_n_frames
    fig = plt.figure()
    ax = fig.add_subplot(111, autoscale_on=False, xlim=(-4,4), ylim=(-2.1, 2.1))
    x = [0, pendulum1_x[i]]
    y = [0, pendulum1_y[i]]
    ax.plot(x,y,lw=3,color='mediumvioletred')
    x1 = [pendulum1_x[i], pendulum2_x[i]]
    y1 = [pendulum1_y[i], pendulum2_y[i]]
    scatter_x.append(pendulum2_x[i])
    scatter_y.append(pendulum2_y[i])
    ax.plot(x1,y1,'o-',lw=3,color='mediumvioletred',markersize=15)
    ax.scatter(scatter_x,scatter_y,lw=0.0005,color='black')
    ax.set_xlabel('$x-Axis$',fontsize=12)
    ax.set_ylabel('$y-Axis$',fontsize=12)
    ax.set_title('Double pendulum simulation (RK4 method)',fontsize=14)
    ax.grid()
    fig.savefig(str(j)+'.png',dpi=600)
    plt.show()

"""
# =============================================================================
# Please check the answers!!!
# =============================================================================
References:

#Implementing the RK4 method in Python 
https://youtu.be/mqoqAovXxWA
by Prof. Niels Walet  

#The formulas for the angular acceleration 
https://www.myphysicslab.com/pendulum/double-pendulum-en.html
    
#Animating the double pendulum (N.B. the implementation used here is different)
https://matplotlib.org/3.2.1/gallery/animation/double_pendulum_sgskip.html
"""






